### PR TITLE
feat: introduce city debt + fix free construction

### PR DESCRIPTION
**Pull request by winex**
### fixes
- fixes free construction when stores are empty (part of #277)

### tl;dr
- cities will now have `city.stats.debt` variable and we'll never loose a cent anymore, ex: when treasury is full or absent _(for any reason)_
- we will now use `payCurrency()` instead of directly inserting currency into treasuries
- main loops now call `pay_to_treasury_all()` frequently which calls `payToTreasury()` (not exported), which actually inserts currency. _these should not be used directly_
- added new more complex `consumeRequired()` function which returns cost **IIF** every item supply is sufficient
`City:: startConstruction()` is now able to check required items cost using this new function
- converted `Constants.CONSTRUCTION_MATERIALS` to nice-and-clean looking dict ;)

naming was choosen in perspective of paying-to-player and to look similar, not something like increaseDebt() and addCurrencyToTreasury()

as i promised - with this you'll able to repay UPC leftovers with just `payCurrency()` to any and every city in #294
what do you think?

### todo
- [x] DONE: resolve "free" `garden` cost issue: 0 => free, nil => "not enough" _-- negative cost value might be used in future_
- [x] DONE: names, gui field text, etc are open subject for discussion
- [ ] reduce logging _-- not yet_
- [x] apply changes after review
_+ something more i forgot about_

### CODE REVIEW

File: city.lua [hunk 1]
```diff
@@ -846,7 +847,7 @@ local function startConstruction(city, buildingConstruction, queueIndex, allowed
             coordinates = Queue.popleft(city[queueIndex])
             if coordinates == nil then
                 -- If there are no more entries left in the queue, then abort
-                return false
+                goto failure
```
Comment (winex) [inline city.lua:850]: i'll rework this `goto` later, too. it just was so simple to log failure reason here by something alike `try-except-finally`

@winex: move logging into sub-function`(..., reason)` and call before `return`

**Comment by winex** [comment]
rebased on latest `main`, removed `goto` statements, resolved cost to be `nil` on requirements failure

looks much better now, marking as ready just to request changes...

File: constants.lua [hunk 2]
```diff
@@ -44,49 +44,29 @@ local CONSTANTS = {
...
+            ["iron-plate"] = 20,
+            ["steel-plate"] = 10,
+            ["small-lamp"] = 2,
+        },
+        simple = {
+            ["stone-brick"] = 10,
+            ["iron-plate"] = 5,
+        },
+        garden = {
+        },
```
Comment (winex) [inline constants.lua:69]: @bahrmichael , as this now looks much easier to read - please recheck this table(dict) as i don't know global structure and initial design about all of this. it might be missing something

note: **all valid building must be listed here** from now on, otherwise `assert()` is raised from `Consumption:: consumeRequired()` to avoid errors before releasing
free construction such as current `garden` must be listed as `{}` empty table
oh, and `"specialBuildings"` was not valid entity name, so i reworked it to (almost full) names

Comment (bahrmichael) [reply constants.lua:69]: > note: all valid building must be listed here from now on, otherwise assert() is raised from Consumption:: consumeRequired() to avoid errors before releasing

Nice!

File: gui.lua [hunk 3]
```diff
@@ -596,6 +596,9 @@ local function addCityOverview(city, anchor)
     -- citizen count
     tbl.add{type = "label", caption = {"", {"tycoon-gui-citizens"}, ": "}}
     tbl.add{type = "label", caption = {"", countCitizens(city)}}
+    -- debt
+    tbl.add{type = "label", caption = {"", {"tycoon-gui-debt"}, ": "}}
```
Comment (bahrmichael) [inline gui.lua:600]: I'm wondering if this will be easy enough to understand for players. Shall we include a description sentence along the lines of "this is how much the city owes you, make space in treasuries"?

**Comment by bahrmichael** [comment]
Those are really great changes and I'll be happy to merge them in after a quick test.

Just had one comment about informing players what "unpaid debt" means for them, but we can also do that post-merge.

File: gui.lua (hunk 3 above)
Comment (winex) [reply gui.lua:600]: additional description would be really helpful, any example of how it's done? i'm eager for settings' `?` round hint

i just can't get proper name for this: "pending payments" or just "unpaid"
first naming was even simpler `Debt:`, but it doesn't mean more for players - "debt of what and to where?" questions will raise

File: locale/en/general.cfg [hunk 4]
```diff
@@ -17,6 +17,7 @@ tycoon-gui-basic-needs-met=Basic needs met
 tycoon-gui-basic-needs=Basic Needs
 tycoon-gui-citizens=Citizens
 tycoon-gui-construction-sites=Construction Sites
+tycoon-gui-debt=Unpaid debt
```
Comment (bahrmichael) [inline locale/en/general.cfg:20]:
```suggestion
tycoon-gui-debt=Pending payouts
tycoon-gui-debt-tooltip=When treasuries are full, the city keeps track of pending payouts. Make space in the city's treasury to release the payout.
```

File: gui.lua [hunk 5]
```diff
@@ -596,6 +596,9 @@ local function addCityOverview(city, anchor)
...
+    tbl.add{type = "label", caption = {"", string.format("%.2f", city.stats.debt)}}
```
Comment (bahrmichael) [inline gui.lua:601]:
```suggestion
    tbl.add{type = "label", caption = {"", string.format("%.2f", city.stats.debt)}, tooltip = {"", {"tycoon-gui-debt-tooltip"}}}
```

File: gui.lua (hunk 3 above)
Comment (bahrmichael) [inline gui.lua:600]:
```suggestion
    tbl.add{type = "label", caption = {"", {"tycoon-gui-debt"}, ": "}, tooltip = {"", {"tycoon-gui-debt-tooltip"}}}
```

**Review by bahrmichael — COMMENTED** [review]
Added some suggestions on using the tooltip

File: locale/en/general.cfg (hunk 4 above)
Comment (bahrmichael) [reply locale/en/general.cfg:20]: What do you think of "pending payouts" instead of "unpaid debt"?

Comment (winex) [reply locale/en/general.cfg:20]: this looks really good!

File: gui.lua (hunk 5 above)
Comment (winex) [reply gui.lua:601]: this `%.2f` format was put here to check that `x.4 + x.5 + x.6` correctly payouts as `1.0` and player never loses fractional coins.

it's simple to change to `%.0f`/`%d` or just `math.floor(city.stats.debt)`. i suppose we don't need `tostring(number)` here, but i'd use it (just in case it's `nil`):
```diff
-    tbl.add{type = "label", caption = {"", string.format("%.2f", city.stats.debt)}}
+    tbl.add{type = "label", caption = {"", tostring(math.floor(city.stats.debt))}}
```

**Comment by winex** [comment]
oh, editing from github was too weird for me.
_do we actually need both tooltips?_

anyway, i must rebase all of these because of conflicts, so i'll add fractional part into second tooltip instead, but integer debt/"pending payout" will be shown

**Comment by bahrmichael** [comment]
> do we actually need both tooltips?

I would suggest adding it for the label and the value, so that it's easier to display the hover and learn more.

**Comment by bahrmichael** [comment]
Code looks overall good to me, let me know when I should merge it :)

**Comment by winex** [comment]
testing this together with #304, looks cool for the quick tests

i'll test (and probably stream) this more tomorrow....
still need to reduce logging a bit, but we need that especially when testing multiple cities/multiple treasuries

**Comment by winex** [comment]
i waited so long for used-bottle-store be merged just to make this commit: 79b286a , haha

**Comment by bahrmichael** [comment]
> i waited so long for used-bottle-store be merged just to make this commit: https://github.com/bahrmichael/factorio-tycoon/commit/79b286acafa74d7b918325db1f37b4a65cf809c9 , haha

Glad that's done now! :D

**Comment by bahrmichael** [comment]
I copied and merge the branch. Thank you very much for this great upgrade!
